x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -366,6 +375,20 @@ public void getRoleForApiKey(Authentication authentication, ActionListener<ApiKe
         }
     }
 
+    public BytesReference getRoleDescriptorsBytesForApiKey(Authentication authentication, boolean limitedBy) {
+        if (authentication.getAuthenticationType() != AuthenticationType.API_KEY) {
+            throw new IllegalStateException("authentication type must be api key but is " + authentication.getAuthenticationType());
+        }
+        final Map<String, Object> metadata = authentication.getMetadata();
+        return (BytesReference) metadata.get(limitedBy ? API_KEY_LIMITED_ROLE_DESCRIPTORS_KEY : API_KEY_ROLE_DESCRIPTORS_KEY);
+    }
+
+    public List<RoleDescriptor> getRoleDescriptorsForApiKey(Authentication authentication, boolean limitedBy) {
```

Comment:
**Short version**: I removed `getRoleDescriptorsForApiKey` and renamed `getRoleDescriptorsBytesForApiKey` to `getApiKeyIdAndBytes`. 

**Longer version**: I agree the names are rather confusing. Initially, I completely removed `getRoleForApiKey` and `List<RoleDescriptor> parseRoleDescriptors(String, Map<String, Object>)` because they are used only for old style `Authentication`. But I had to add them back when I realised there were BWC issues. Anyway, we can and should remove them in 9.0. 

Unfortunately, it is not feasible to merge `getRoleForApiKey` and `getRoleDescriptorsForApiKey` even though they both return something related to role descriptors. The old `getRoleForApiKey` returns a `ApiKeyRoleDescriptors` object, which internally has two `List<RoleDescriptors>`, one for from-role and the other for limited-by. But this would be non-ideal for the new `getRoleDescriptorsForApiKey` which returns either from-role-descriptors or limited-by-role-descriptors. By returning them separately, we can cache and lookup them separately. That is, if multiple keys are created by the same user, their limited-by roles are only computed once and cached for all other keys. The reverse is also true (though maybe less likely), if multiple keys happen to have the from-role-descriptors even when they are created by different users, we only need compute the from-roles once. 

So there is no shared code path between `getRoleForApiKey` and `getRoleDescriptorsForApiKey` they work like the follows:
* Old Authentication -> getRoleForApiKey -> parseRoleDescriptors(String, Map<String, Object>)
* New Authentication -> getRoleDescriptorsBytesForApiKey -> if not cached -> getRoleDescriptorsForApiKey -> parseRoleDescriptors(String, ByteReference)

I intentionally kept them separate so the old code path can be easily removed when the time comes.

With above being said, I do agree that the names are confusing. So I renamed `getRoleDescriptorsBytesForApiKey` to `getApiKeyIdAndBytes` and removed `getRoleDescriptorsForApiKey` by a direct call to `parseRoleDescriptors(String, ByteReference)` in the caller.

Reply:
If I understand correctly, `getRoleForApiKey` will only do the right thing for old style authentication.
If so, can we:
1. Add that to the javadoc
2. Add an `assert authentication.getVersion().before(Version.V7_9_0) : "This method only applies to authentication objects created before v7.9.0"`;

Reply:
These are both good ideas. Will update accordingly. Thanks!